My plan: Add checklist progress banner

sirreal opened the pull request:
Prepare checklist UI for the My Plan page:

***The checklist is not visible, #12404 will make the checklist functional***

![Screen Shot 2019-05-08 at 12 39 11](https://user-images.githubusercontent.com/841763/57369644-54973080-718e-11e9-91c6-7881f1f3cbea.png)

The data layer is in place, but no API calls are made. Endpoints in Jetpack and WordPress.com need to be prepared. Watch #12404 for work there.

Part of https://github.com/Automattic/wp-calypso/issues/32565

### Testing
- Builds 👍 
- No changes at `/wp-admin/admin.php?page=jetpack#/my-plan` 👍 
- ~Does the link to Calypso work and direct you to the My Plan page in Calypso?~

sirreal committed:
Add initial pass at progress banner

sirreal committed:
Add state for checklists

sirreal committed:
Hook up card to state

sirreal committed:
Styles

sirreal committed:
Make the button work

sirreal committed:
Fix colors spacing

simison commented on _inc/client/scss/style.scss:
```diff
@@ -13,6 +13,7 @@
 
 // Variables
 @import 'variables/colors';
+@import '../../../node_modules/@automattic/calypso-color-schemes/src/__color-studio/color-variables.scss';
```
Remember to add `@automattic/calypso-color-schemes` to direct dependencies (it's now pulled in via `calypso-build`)

simison replied:
Anything here that might end up unintentionally overwriting variables from `variables/colors`?

sirreal replied:
Good point! I'll change the order so even if there are conflicts, the variables in `variables/colors` has precedence.

sirreal committed:
Add calypso-color-schemes dep

sirreal committed:
Order sass imports so local colors have precedence

sirreal committed:
Improve button styling

simison commented on .eslintrc.js:
```diff
@@ -100,6 +100,12 @@ module.exports = {
+		{
+			files: [ '_inc/client/**/*' ],
+			globals: {
+				__AppFeatures__: 'readonly',
```
Is this "globals" as in `window` or globals for the app bundle? Should we prefix this with `Jetpack`

sirreal committed:
Add jetpack prefix to progress bar animation

sirreal commented on _inc/client/my-plan/checklist-progress-card/progress-bar/screen-reader-text/style.scss:
```diff
@@ -0,0 +1,26 @@
+// Text meant only for screen readers
+.screen-reader-text { // TODO add _accessibility.scss
```
Incredible…

@lezama are you planning on getting back to this? 😂 

1075-gh-calypso-pre-oss

I think I can drop the todo 😅

sirreal committed:
Remove TODO comment

sirreal committed:
Use site slug over site ID

sirreal commented on _inc/client/my-plan/checklist-progress-card/index.js:
```diff
@@ -0,0 +1,73 @@
+import { getTasks } from 'state/checklist/selectors';
+
+// Style imports are unused
+// import './style.scss';
```
Yeah… I was dreaming of a day when the build system uses webpack-based sass… I'll remove them.

simison replied:
These lines should be cleaned out?

sirreal committed:
Remove style imports

sirreal commented on webpack.config.js:
```diff
@@ -6,6 +6,8 @@ const WordPressExternalDependenciesPlugin = require( '@automattic/wordpress-exte
 const NODE_ENV = process.env.NODE_ENV || 'development';
 const devMode = process.env.NODE_ENV !== 'production';
 
+const enabledFeatures = process.env.ENABLE_FEATURES ? process.env.ENABLE_FEATURES.split( ',' ) : [];
```
I have reservations about making this official by way of documenting it… I was planing to remove it completely when this feature is landed unless folks really want something like this.

If we need something like this in Jetpack, we might need to give it a bit more though 🙂

sirreal replied:
Dropped all the feature flag stuff. The checklist is hidden now because it doesn't fetch any data.

sirreal commented:
Rebased to fix conflict

jeherve commented on _inc/client/my-plan/checklist-progress-card/progress-bar/screen-reader-text/index.js:
```diff
@@ -0,0 +1,8 @@
+/**
```
Should this component be moved up a bit? I can imagine it being useful in other parts of the dashboard in the future, so having it somewhere less hidden could be useful?

sirreal replied:
I added things deeper in the directory structure thinking that it might be easier to replace them with a shared component library when and if that appears.

That may be naive and I'm not sure whether the premise holds. Happy to move this up if you prefer.

jeherve replied:
That makes sense. I am just afraid someone may think the same and add it somewhere else later on, without looking for it first. :)

ockham replied:
> a shared component library when and if that appears.

xrefing https://github.com/Automattic/wp-calypso/pull/32869#issuecomment-493029184

sirreal replied:
This component is super simple and depends on core styling. I've moved it to shared `components`.

jeherve commented on yarn.lock:
```diff
@@ -41,7 +41,7 @@
     webpack-filter-warnings-plugin "1.2.1"
     webpack-rtl-plugin "1.8.0"
 
-"@automattic/calypso-color-schemes@^1.0.0":
```
I don't really understand what triggered that change?

sirreal replied:
I added the direct dependency.

This lines says that both required versions, (with `^` from a package and without as the direct dependency) resolve to the following lines.

jeherve commented on webpack.config.js:
```diff
@@ -66,6 +68,11 @@ const webpackConfig = {
 			// only be used in development
 			'process.env.NODE_ENV': JSON.stringify( NODE_ENV ),
+
+			// Feature flags!
```
How about relying on the Dev Tools Card for now to test changes to the checklist status?
https://github.com/Automattic/jetpack/blob/master/_inc/client/components/dev-card/index.jsx

jeherve commented on _inc/client/my-plan/checklist-progress-card/index.js:
```diff
@@ -0,0 +1,59 @@
+						<ProgressBar compact canGoBackwards total={ total } value={ completed } />
+					</div>
+					<div className="checklist__header-secondary">
+						<Button compact primary href={ `https://wordpress.com/plans/my-plan/${ siteSlug }` }>
```
You'll probably need to add onClick tracking here.

sirreal replied:
@jeffgolenski ☝️ 

I'll open an issue to handle that in a follow-up

jeherve replied:
#12335 12335

jeherve commented on _inc/client/my-plan/checklist-progress-card/style.scss:
```diff
@@ -0,0 +1,94 @@
+	.dops-button.is-primary {
+		margin-right: 0;
+		background: $green-primary;
+		border-color: $green-secondary;
```
Yes, it's a bit weird to have blue and green CTAs on the same page I think:

![image](https://user-images.githubusercontent.com/426388/57454174-82e93e80-7268-11e9-9c89-64156a4f6765.png)


Maybe that new button should be blue?

Since we are comparing buttons, maybe we should look at not uppercasing the text in that new button?

sirreal replied:
@jeffgolenski ☝️

jeffgolenski replied:
If we have time here, I think it would be prudent to build some logic in. The color is less important. What's important is that we only have one primary CTA show at any given time. 

Now that we're implementing the checklist I'd argue the only primary CTA should be the "continue jetpack setup" button. Let's make it blue for now and make all other CTAs white.

Once the jetpack checklist has been completed, that button should disappear completely. then we can make that "view settings" button the primary CTA.

sirreal replied:
I've opened a simple follow-up, I don't want to get into all of these changes in this PR. Feel free to add details @jeherve and @jeffgolenski 

#12405

jeherve commented on _inc/client/my-plan/my-plan-header.jsx:
```diff
@@ -135,7 +136,12 @@ class MyPlanHeader extends React.Component {
 		}
-		return <div>{ planCard }</div>;
+		return (
+			<>
```
Should we give this container a name?

sirreal replied:
This is the [short syntax for React `Fragment`s ](https://reactjs.org/docs/fragments.html#short-syntax). It's equivalent to `<Fragment>`.

jeherve replied:
Good to know, thanks!

jeherve commented on _inc/client/state/action-types.js:
```diff
@@ -142,3 +142,6 @@ export const JETPACK_SITE_VERIFY_GOOGLE_REQUEST = 'JETPACK_SITE_VERIFY_GOOGLE_RE
 	'JETPACK_SITE_VERIFY_GOOGLE_REQUEST_SUCCESS';
 export const JETPACK_SITE_VERIFY_GOOGLE_REQUEST_FAIL = 'JETPACK_SITE_VERIFY_GOOGLE_REQUEST_FAIL';
+
+export const SITE_CHECKLIST_RECEIVE = 'SITE_CHECKLIST_RECEIVE';
```
Should we include `PLAN` in the constant names, just in case we add more checklists in the future?

sirreal replied:
🤔 I can make the change but I'm not sure there will be any conflict or that we need to address it preemptively.

In implementing this, I was thinking about receiving checklist data from the API. There's one checklist endpoint and it returns a bunch of data, one part  of it is a tasks object.

If we to want different names for these Redux action types in the future, it's should be pretty straightforward to change the name in a few files.

jeherve replied:
That works for me 👍

jeherve commented on _inc/client/my-plan/checklist-progress-card/index.js:
```diff
@@ -0,0 +1,59 @@
+import Button from 'components/button';
+import Card from 'components/card';
+import ProgressBar from './progress-bar';
+import QueryChecklistProgress from 'components/data/query-checklist-progress';
```
Will `QueryChecklistProgress` return a specific thing when the list is already complete, so we can manipulate (or even hide) the card accordingly, or should that be handled by each component by comparing the number of steps completed to the total?

![image](https://user-images.githubusercontent.com/426388/57454928-43bbed00-726a-11e9-8f32-4d36e0f713a8.png)

sirreal replied:
`QueryChecklistProgress` is a data component that will make sure that data we're interested in is fetched.

Components typically get the data by calling selectors. The most basic selectors just find the data in state and return it, but what you mention is a job that would fit well in a selector: transform checklist data into completion data.

In this PR I've done that inline:

https://github.com/Automattic/jetpack/blob/21108a1084b6d8b68bcb5a466fa9daa7d671934a/_inc/client/my-plan/checklist-progress-card/index.js#L49-L52

That could certainly (and arguably should) be moved (and optimized) in a dedicated selector.

jeherve replied:
> moved (and optimized) in a dedicated selector.

Right, that's kinda what I had in mind; I think that's information anyone building a checklist will need, so we might as well make it available to everyone.

sirreal replied:
Implemented the selector.

jeherve reviewed (dismissed):
Just a few questions about the implementation.

simison commented on package.json:
```diff
@@ -148,6 +148,7 @@
 	},
 	"devDependencies": {
 		"@automattic/calypso-build": "1.0.0",
+		"@automattic/calypso-color-schemes": "1.0.0",
```
Isn't this now `dependencies` since it's used in the code, not just `devDependencies`?

sirreal replied:
It's used at the same time as `calypso-build` 🤷‍♂ 

I think the distinction and separation is nonsensical in this package.json.

What's the production app? Jetpack the plugin? That doesn't rely on any dependencies (node_modules) at run time.

Or is the app this package.json describes a Jetpack builder? In that case everything is a dependency and what are devDependencies for? (I lean towards this thinking).

I think the distinction is mostly useful for distributed modules, which this is not.

For cleanliness and consistency, I'd put all of the dependencies in this package.json into `dependencies` and call it a day. But I don't want to do that in this PR and I don't think it makes any difference where the dependency is for this app.

simison replied:
>Or is the app this package.json describes a Jetpack builder? In that case everything is a dependency and what are devDependencies for? (I lean towards this thinking).

Oh yeah I'm in this camp totally too. :-( I use the distinction myself just to communicate how something is used in the repo and practically everything always gets installed anyway.

No strong opinions where it is here since it doesn't indeed have difference in practise. :-)

simison reviewed (dismissed):
Mobile view is broken:

<img src="https://user-images.githubusercontent.com/87168/57696436-6d558980-7659-11e9-9fdf-de0f46dad584.png" alt="" width="375"/>

sirreal committed:
Implement API for checklist

jeffgolenski committed:
checklist progress card: adjusting mobile styles

checklist progress card: adjusting mobile styles for smaller screens

jeffgolenski commented:
@simison @sirreal I've adjusted the mobile view so it's not broken. 

![Screen Shot 2019-05-14 at 2 50 52 PM](https://user-images.githubusercontent.com/214813/57724166-2915b980-7658-11e9-9bc1-9b625e855056.png)

sirreal committed:
Hide the card when empty

sirreal committed:
Add completed/total selector

sirreal commented:
I polished this a bit and addressed some issues. I also added an early implementation of the endpoint, but that may require some more work.

I can pull the data bits back out, without the data the visual changes are hidden.

simison commented on _inc/client/my-plan/checklist-progress-card/progress-bar/index.js:
```diff
@@ -0,0 +1,77 @@
+		const styles = { width: this.getCompletionPercentage() + '%' };
+		if ( this.props.color ) {
+			styles.backgroundColor = this.props.color;
+		}
```
What do you think if we'd drop this inline style prop from this component? It's used just once in Calypso and I'd like to remove it there, too.

sirreal replied:
A fair amount of the functionality is unused. Do you think I should strip this down to be a custom progress bar just for our case?

I'm not sure whether it's best to make it easy to replace with the Calypso version or specialized for this use.

simison replied:
True, prolly not worth it. 👍

simison commented on _inc/client/my-plan/checklist-progress-card/progress-bar/screen-reader-text/style.scss:
```diff
@@ -0,0 +1,26 @@
+// Text meant only for screen readers
+.screen-reader-text {
```
Core has this style already for wp-admin — perhaps we don't need it at all?

https://github.com/WordPress/WordPress/blob/80aee4ae87343ea3990314c453793d334beb8ebb/wp-admin/css/common.css#L120-L135
https://github.com/WordPress/WordPress/blob/80aee4ae87343ea3990314c453793d334beb8ebb/wp-admin/css/wp-admin.css#L1

sirreal replied:
Dropped the styles.

simison commented on _inc/lib/class.core-rest-api-endpoints.php:
```diff
@@ -477,6 +490,32 @@ public static function get_plans( $request ) {
+			array(
+				'method'  => 'GET',
+				'headers' => array(
+					'X-Forwarded-For' => Jetpack::current_user_ip( true ),
```
Just curious; is this header needed? I see other functions in this file have it but e.g. this doesn't:

https://github.com/Automattic/jetpack/blob/8f05c29dabbfe7161a3f932097bc1663db78d035/_inc/lib/core-api/wpcom-endpoints/memberships.php#L111-L123

sirreal replied:
I'm not sure, this was just working from similar request.

Data bits have moved to a follow-up PR: https://github.com/Automattic/jetpack/pull/12404

sirreal committed:
Remove checklist API pieces

Checklist API requires WPCOM-side work.
Disable and remove the data fetching for now.

sirreal committed:
Remove screen-reader-text styles (use core)

sirreal committed:
Move to ScreenReaderText to shared components

sirreal commented:
I believe all review points have been addressed. This PR now contains no visible changes for users.

The follow-up #12404 which adds necessary data will make the checklist functional and there will be another opportunity for review.

I'd appreciate another round of reviews at this time. Testing should ensure that the dashboard is functionally unchanged but the code can still be reviewed.

simison commented on _inc/client/my-plan/checklist-progress-card/progress-bar/style.scss:
```diff
@@ -0,0 +1,52 @@
+.progress-bar {
```
This might be a little too generic class name to have at global scope.

sirreal replied:
Good point. I rolled all of this into a specific component and specific styles now.

simison reviewed (dismissed):
Looking fabulous and indeed doesn't show up in dashboard just yet.

Left just a small nit about generic CSS class name, feel free to leave out if you think it's not a concern.

Good to ship!

sirreal committed:
Simplify and make progress-bar specific

jeherve reviewed (approved):
This looks good to me. 👍

sirreal commented:
r191989-wpcom

sirreal commented:
I've closed https://github.com/Automattic/jetpack/pull/12404 with no plans to continue.

It's possible that several files added here are completely unused and can be safely removed. It should be relatively straightforward cleanup.

FYI: @Automattic/jetpack-crew

jeherve commented:
> It's possible that several files added here are completely unused and can be safely removed. It should be relatively straightforward cleanup.

Logged in #14620.

jeffgolenski commented:
Thanks for keeping an eye on this y'all.

Just cc'ing the design team so they know it exists. We may revisit in the future: @keoshi @crunnells  @eeeeevon13 @folletto
